core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -214,41 +222,44 @@ public void dataRehashed(DataRehashedEvent<K, V> event) {
    private Stream<Original> getRehashStream(CacheSet<Original> cacheEntrySet, Object requestId,
          SegmentListener listener, boolean parallelStream, boolean entryStream, IntSet segments,
          Set<K> keysToInclude, Set<K> keysToExclude) {
-      LocalizedCacheTopology topology = dm.getCacheTopology();
-      if (trace) {
-         log.tracef("Topology for supplier is %s for id %s", topology, requestId);
-      }
-      ConsistentHash readCH = topology.getCurrentCH();
-      ConsistentHash pendingCH = topology.getPendingCH();
-      if (pendingCH != null) {
-         IntSet lostSegments = IntSets.mutableEmptySet(topology.getCurrentCH().getNumSegments());
-         PrimitiveIterator.OfInt iterator = segments.iterator();
-         while (iterator.hasNext()) {
-            int segment = iterator.nextInt();
-            // If the segment is not owned by both CHs we can't use it during rehash
-            if (!pendingCH.isSegmentLocalToNode(localAddress, segment)
-                    || !readCH.isSegmentLocalToNode(localAddress, segment)) {
-               iterator.remove();
-               lostSegments.set(segment);
-            }
+      // Don't even bother checking topology if listener is null - this means REPL
+      if (listener != null) {
```

Comment:
This method is getting kind of big, the listener part could be a separate method.

Reply:
Okay